### DISCUSSION TITLE
Fork the repo but when i installed filament from my forked repo i get a Error

### DISCUSSION THREAD

**Discussion by sohailahmad07** [filamentphp/filament#14415] [General]
https://github.com/filamentphp/filament/issues/14394

Fix Pull Request 
https://github.com/filamentphp/filament/compare/3.x...sohailahmad07:filament:add_disable_grammarly

Hi to fix the above simple issue i fork the repo and fix it but before i make a pull request i want to test it locally so i installed a fresh laravel project and in my composer.json i add
```
  "require-dev": {
        "filament/filament": "dev-add_disable_grammarly",
}
 "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/sohailahmad07/filament"
        }
    ]
```

but when I run composer install i get the error 
```
Generating optimized autoload files
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
Script Illuminate\Foundation\ComposerScripts::postAutoloadDump handling the post-autoload-dump event terminated with an exception

In autoload_real.php line 41:
                                                                                                                                                                     
  require(/mnt/DevOps/devops/Open-Source/Testing-Filament/vendor/composer/../filament/filament/tests/helpers.php): Failed to open stream: No such file or directory  
                                                                                                                                                                     
install [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--download-only] [--dev] [--no-suggest] [--no-dev] [--no-autoloader] [--no-progress] [--no-install] [--audit] [--audit-format AUDIT-FORMAT] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] ...

```

and i a positive i did not create this issue so i create a new branch for 3.x (default) and with out any change i push that branch and when i tried to installed form that brand i get the same error

**Comment by corey-si** [comment]
I seem to be getting this same issue when trying to do it from a private packagist repo from a forked version, is there something more to this?

**Reply by sohailahmad07** [reply]
i thought i was the only one let me know if you find a solution

**Comment by shopapps** [comment]
same here, not sure what i'm doing wrong...

tried editing composer.json
```json
...
"repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/shopapps/filament"
        }
    ],
    
...
    
"filament/filament": "dev-csp-nonce-support as v3.2.117",
```

when running `composer update` I'm seeing lots of errors along the lines of
```bash
composer update        
 
Loading composer repositories with package information                                                                                                                                                           Updating dependencies                                 
Lock file operations: 0 installs, 1 update, 0 removals
  - Upgrading filament/filament (v3.2.108 => dev-csp-nonce-support 1a4feeb)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Downloading filament/filament (dev-csp-nonce-support 1a4feeb)
  - Upgrading filament/filament (v3.2.108 => dev-csp-nonce-support 1a4feeb): Extracting archive
Generating optimized autoload files

Warning: Ambiguous class resolution, "Filament\Actions\Concerns\HasWizard" was found in both "/Users/xxxxx/projects/xxxx/vendor/filament/actions/src/Concerns/HasWizard.php" and "/Users/xxxx/projects/xxxx/vendor/filament/filament/packages/actions/src/Concerns/HasWizard.php", the first will be used.
```

**Comment by f-liva** [comment]
After banging my head for a long time, I finally figured out how to use custom forks for Filament, which they would have been nice to document somewhere, but instead they don't.

---

# 🛠️ How to use a forked Filament package (e.g. `filament/forms`)

## 🔎 The problem

Filament is maintained as a **monorepo** (`filamentphp/filament`) but is published on Packagist using **subtree splits**.
This means the actual Composer packages are separate:

* `filament/forms`
* `filament/tables`
* `filament/actions`
* …etc.

If you want to modify the code, forking the monorepo is not enough.
You must **fork the specific package** you want to change (e.g. `filamentphp/forms`).

---

## ✅ The solution step by step

### 1. Fork the correct package

If you need to modify **Forms**, fork [`filamentphp/forms`](https://github.com/filamentphp/forms).
For other modules, fork the corresponding subtree split.

---

### 2. Add your fork in `repositories`

In your project’s `composer.json`:

```json
"repositories": [
    {
        "type": "vcs",
        "url": "git@github.com:your-username/filament-forms.git"
    }
]
```

---

### 3. Add the **same tag** as the official release

Filament packages depend on **exact versions** (e.g. `v4.0.16`).
If your fork doesn’t have this tag, Composer can’t resolve dependencies.

In your fork, create and push the tag:

```bash
git checkout 4.x   # or the appropriate branch
git tag v4.0.16    # must match the official release tag
git push origin v4.0.16
```

---

### 4. Require the package in your project

Update your `composer.json`:

```json
"require": {
    "filament/filament": "^4.0",
    "filament/forms": "v4.0.16"
}
```

👉 No need for aliases like `dev-main as 4.0.16`.
With the tag, Composer sees your fork as a drop-in replacement.

---

### 5. Update dependencies

Clear Composer cache and update:

```bash
composer clear-cache
composer update filament/forms -W
```

The `-W` option allows updating other packages locked to the same dependency.

---

## 🎯 Why is the tag required?

* Filament (`filament/filament`) and many third-party packages require **specific tagged versions** (e.g. `v4.0.16`).
* Without the tag, Composer only sees `dev-main` or `4.x-dev`, which **don’t satisfy the constraint** → resolution error.
* With the correct tag, your fork is indistinguishable from the official package, so everything works seamlessly.

---

## 🚀 TL;DR

1. Fork the package you need (e.g. `filament/forms`).
2. Add the same version tag as the official release.
3. Register your fork in `repositories`.
4. Require the tagged version in `composer.json`.
5. Run `composer update`.

Result: you’re using your forked Filament package without breaking compatibility with the ecosystem.
